Joy sql normalization (#269)
arcus/education_modules: database_normalization/database_normalization.md

<!--
author:   Joy Payton
email:    paytonk@chop.edu
version:  1.0.0
module_template_version: 3.0.0
language: en
narrator: US English Female
title: Database Normalization
comment: Learn about the concept of normalization and why it's important for organizing complicated data in relational databases.
long_description: Usually, data in a relational database like SQL is organized into multiple interrelated tables with as little data repetition as possible. This concept can be useful to apply in other areas as well, such as organizing data in .csvs or in data frames in R or Python.  This module teaches underlying data considerations and explains how data can be efficiently organized by introducing the concepts of one-to-many data relationships and normalization.
estimated_time: 40 minutes

@learning_objectives  
After completion of this module, learners will be able to:

- Explain the significance of "one to many" data relationships and how these relationships affect data organization
- Describe how a normalized database is typically organized

> - Explain how data can be linked between tables and define "primary keys" and "foreign keys"

Reviewer: Question: are the terms "primary key" and "foreign key" specific to SQL or are they general information theory terms?

Author: I'd say they are "relational database" terms, e.g. Microsoft Access uses them, SQL uses them, and people use them "off label" even when they don't apply (e.g. Google BigQuery doesn't have foreign keys but we talk about "using xyz as a foreign key."

Reviewer: Great! Good to know. Any chance this could fit in as a "behind the scenes" or "cool fact" or "history" box in the Key Vocabulary section?

Author: ooh I like it!  Will do!

* One encounter may include multiple orders for different procedures and medications.
* One medication order can give rise to multiple medication administrations.

**Retail website one-to-many examples**

* One customer may have several associated addresses for shipping, billing, or both.
* One customer will probably have multiple orders.
* Each order will have one or many associated items.

One-to-many data relationships (one X is related to more than one Y) can be very hard to accommodate in a single table with all the data.  Imagine having a table of order information from an online website, in which a single row contained every item included in that order. 

It would be impossible to know ahead of time how many columns would be needed.  What if we provided enough columns for 100 items in an order to start?  Then, maybe someone orders 110 items.  What do we do?  Also, most orders are much smaller, which means we have many empty cells with no data in them.  It's not an efficient way to store data.  

### Quiz: One-to-Many 

Which of the following relationships are likely to be one-to-many?  Select all the correct answers.

[[X]] Gym member to workout

> [[ ]] Street address to corresponding legislative district

Reviewer: ```suggestion
[[ ]] Street address to latitude and longitude coordinates
```
If you want X:Y to be one to one and X is "street address," the only thing I can think of for Y is its lat/long. Anything regarding a political boundary either has to get way too specific to actually be one-to-one e.g. "state legislative district in a US state with a unicameral legislature" or doesn't work in parts of the world where borders are disputed, e.g. "country."

Author: so much better!  agree

[[X]] Patient to medical provider
[[X]] Medical provider to patient
[[ ]] Baseball team to mascot
[[?]] There are multiple correct answers!
*********

<div class = "answer">

> We hope that any gym member will have more than one recorded workout, so that relationship is one-to-many.  However, a particular street address will only have one legislative district representing residents at that address.  That's not a one-to-many relationship.  A patient can have multiple providers, so that's one-to-many, and providers can have many patients, so that's also one-to-many.  In fact you could more accurately call this relationship many-to-many!  Finally, a baseball team (we argue) should have only one mascot, so that's not a one-to-many relationship.

Author: ```suggestion
We hope that any gym member will have more than one recorded workout, so that relationship is one-to-many.  However, a particular street address will only be related to a single latitude and longitude. That's not a one-to-many relationship.  A patient can have multiple providers, so that's one-to-many, and providers can have many patients, so that's also one-to-many.  In fact you could more accurately call this relationship many-to-many!  Finally, a baseball team (we argue) should have only one mascot, so that's not a one-to-many relationship.
```

**Table 1: depression\_scale**

<!-- data-type="none" class="tight-table" style="font-size:80%"-->
| subj_id  | date  | dep_q1  | dep_q2   | dep_q3  | dep_q4  | dep_total   |
| :--------- | :--------- | :--------- | :--------- | :--------- | :--------- | :--------- |
| 11234   | 2021-05-15   | 3    | 3  | 2    | 4    | 12    |
| 86234   | 2021-06-01   | 4    | 4  | 3    | 4    | 15    |
| 32660   | 2021-06-10   | 1    | 1  | 2    | 1    | 5   |
| 86234   | 2022-01-13   | 2    | 2  | 1    | 3    | 8    |
| 41356   | 2022-02-10   | 1    | 3  | 2    | 3    | 10   |

**Table 2: subject\_address**

<!-- data-type="none" class="tight-table" style="font-size:80%" -->
| subj_id  | street_address  | city  | state   | zip  | date_start  | date_end   |
| :--------- | :--------- | :--------- | :--------- | :--------- | :--------- | :--------- |
| 11234   | 123 Main Street   | Smithtown    | PA  | 19000    | 2022-01-01   | `NULL`    |
| 11234   | 123 Oak Lane   | Old Towne    | PA  | 18000   | 2000-01-01    | 2021-12-31    |
| 93452   | 123 Green Blvd  | Kirby    | TN  | 37000    | 2020-05-01    | `NULL`   |

> Notice the one-to-many aspects of this data.  The subject with id 86234 has two different administrations of the depression inventory, on different dates.  The subject with id 11234 has two different addresses for two different date ranges.

Reviewer: ```suggestion
Notice the one-to-many aspects of this data.  The subject with subj_id 86234 has two different administrations of the depression inventory, on different dates.  The subject with subj_id 11234 has two different addresses for two different date ranges.
```

> **Column Conundrums**

Reviewer: ```suggestion
**Column Conundrums**
-----
```
LiaScript subheadings

**One Table?  Not a Good Idea!**

In short, if we try to force all the data we care about into a single table, we have a problem of rapidly multiplying columns and data repetition in rows.  This makes maintaining, cleaning, correcting, and understanding data very tricky.

We're much better off collecting various groups of facts about our research subjects into distinct, separate tables, like a "demographics" table, a "depression" table, an "anxiety" table, an "address" table, and so on, including a "subj\_id" column to link rows back to the subject in question.  This realization, of how we can separate data into concepts that each get a table, leads us to the idea of **normalization**.

## Normalization

> When data is normalized, we organize data to reduce the possibility of needless data duplication and empty cells.  Normalization is a complex information science topic and has different degrees of completeness, so here we'll talk about what normalization **looks like** more than what it **is**.  

Reviewer: What does "degrees of completeness" mean in this setting? To my ears that phrase sounds like something that behind the scenes has an extremely technical definition, but I actually have no idea.

Author: I mean yes there are "forms" like first, second, third, fourth form normal... I don't think ppl need to know the details about this.  What about:

```suggestion
When data is normalized, we organize data to reduce the possibility of needless data duplication and empty cells. 
Normalization is a complex information science topic and there are various levels of normalization that can be performed.  Here, we won't go into details about various forms of normalization, rather, we'll talk about what normalization  **looks like** more than what it **is**.  
```

Reviewer: I like that a lot better. My mathematician brain gets antsy when "completeness" is mentioned but not defined.

<div class = "behind-the-scenes">
<b style="color: rgb(var(--color-highlight));">Behind the scenes</b><br>
In this module, we're discussing a kind of normalization called "third normal form", often abbreviated 3NF.  The theory behind normalization has its own challenging jargon, like "cardinality", "atomicity", "decomposition", etc.  If you get excited by database optimization and information science, you might enjoy a deep dive into normalization.  But for most of us, we only need to develop a casual understanding of how normalization affects our databases. 
</div>

> **Entities: What Kinds of Things Exist?**

Reviewer: ```suggestion
**Entities: What Kinds of Things Exist?**
-----
```

In a normalized database, we separate data into tables representing different logical **entities**, like "address", "procedure", "medication", "device", etc., and give each of these entities an identifier (like "device\_id" or "medication\_id") to identify a particular instance of the entity.  

> This id number can also be a **primary key** if it will only appear once in the entity table (say, for an item identifier that you don't want to be duplicated).  Imagine a "medication" table with rows that include a medication id and information about that medication (like its brand name, generic name, and whether it requires a prescription).  Or think of an "item" table which includes an item id, along with details about that item, like the manufacturer and description.  Each row is unique, and the id as a primary key won't repeat.  If you're using a SQL database, it's very helpful to know that SQL enforces this and won't let you add a new row in a table with a primary key that already exists.

Author: You're right haha ID it is..

> This id number can also be a **primary key** if it will only appear once in the entity table (say, for an item identifier that you don't want to be duplicated).  Imagine a "medication" table with rows that include a medication id and information about that medication (like its brand name, generic name, and whether it requires a prescription).  Or think of an "item" table which includes an item id, along with details about that item, like the manufacturer and description.  Each row is unique, and the id as a primary key won't repeat.  If you're using a SQL database, it's very helpful to know that SQL enforces this and won't let you add a new row in a table with a primary key that already exists.

Author: ```suggestion
This ID number can also be a **primary key** if it will only appear once in the entity table (say, for an item identifier that you don't want to be duplicated).  Imagine a "medication" table with rows that include a medication ID and information about that medication (like its brand name, generic name, and whether it requires a prescription).  Or think of an "item" table which includes an item ID, along with details about that item, like the manufacturer and description.  Each row is unique, and the ID as a primary key won't repeat.  If you're using a SQL database, it's very helpful to know that SQL enforces this and won't let you add a new row in a table with a primary key that already exists.
```

> We tend to only include in these entity tables the data elements that have a single value that is relatively stable (like medication generic name, device bar code, patient date of birth, or medical procedure billing code).  We may exclude things from entity tables that could have multiple values or change frequently (for example, we would probably exclude patient insurance provider from the patient table).

Reviewer: ```suggestion
We tend to only include in these entity tables the data elements that have a single value that is relatively stable (like medication generic name, device bar code, patient date of birth, or medical procedure billing code).  We may exclude things from entity tables that could have multiple values or change frequently (for example, in the United States we would probably exclude patient insurance provider from the patient table).
```

> **Interactions: How do Entities Interact?**

Reviewer: ```suggestion
**Interactions: How do Entities Interact?**
-----
```

> Then, we make tables that represent **interactions** between entities.  For example, interactions between patients and medications might be stored in a "prescriptions" or "medication\_orders" table. Interactions between addresses and devices might be in a "device\_shipment" table.  These record relationships by including the id numbers of the entities involved, along with other data like the date of the interaction and other details.  A table called "medication\_orders", for example, will include a field for the patient id, a field for the medication id, a field for the prescribing provider id, the date and time of the medication order, the dose and duration of the order, the mode of medication delivery, and other important details such as any notes made by the prescriber.

Author: ```suggestion
Then, we make tables that represent **interactions** between entities.  For example, interactions between patients and medications might be stored in a "prescriptions" or "medication\_orders" table. Interactions between addresses and devices might be in a "device\_shipment" table.  These record relationships by including the ID numbers of the entities involved, along with other data like the date of the interaction and other details.  A table called "medication\_orders", for example, will include a field for the patient ID, a field for the medication ID, a field for the prescribing provider ID, the date and time of the medication order, the dose and duration of the order, the mode of medication delivery, and other important details such as any notes made by the prescriber.
```

Whether these tables of interactions have primary keys and their own identifiers depends, but they often will.  For example, a "device\_shipment" table might have a primary key called "device\_shipment\_id".

### Why Bother With Normalization?

To show why organizing data in a normalized way is helpful, let's use an example.

> Which of the following is more efficient to correct, if we want to change "orane juice" to "orange juice"? Is it easier to correct with the data stored in one table, or with the data in two tables, where items are given an id number?

Reviewer: ```suggestion
Consider the following example of a typo in a database. Is it easier to correct "orane juice" to "orange juice" with the data stored in one table, or with the data in two tables, where items are given an id number?
```

> **One Table Option**

Reviewer: ```suggestion
**One Table Option**
-----
```

**orders**

<!-- data-type="none" class="tight-table" -->
| order_num   | item_1   | item_2  | ...  | item_50  |
| :--------- | :--------- | :--------- | :--------- | :--------- |
| 23125    | orane juice     | pistachios     | ...    | `NULL`    |
| 41320    | peanut butter    | plain bagels     | ...    | orane juice    |
| 53011    | napkins    | distilled water     | ...    | `NULL`    |
| 14123    | pistachios    | orane juice     | ...    | peanut butter    |

In the one table option, we have to look for the phrase "orane juice" anywhere in 50 columns across all rows (consider that there could be thousands of rows).

> **Two Table Option**

Reviewer: ```suggestion
**Two Table Option**
-----
```

**items**

> *Note: here we are using the "item\_id" field as a primary key as well, which is why we include (PK) in the table header.  Often, database tools will do something similar and display a special symbol (like <span class="fa-solid fa-key"></span>) or the letters "PK" to draw your attention to the fact that a particular field is a primary key.* 

Reviewer: Maybe combine this note and the note under order\_items into a single "important" highlight box?

Author: I like it, that's a good use of that box!

<!-- data-type="none" class="tight-table" -->
| item_id (PK)  | item_name  |
| :--------- | :--------- |
| 15 | distilled water |
| 178 | napkins |
| 210 | orane juice |
| 97 | peanut butter |
| 108 | pistachios  |
| 233 | plain bagels  |

**order_items**

*Note: here we want to allow the "order\_id" to repeat as many times as necessary to include all items, so it's not a primary key.* 

<!-- data-type="none" class="tight-table" -->
| order_id | item_id |
| :--------- | :--------- |
| 23125 | 210 |
| 23125 | 108 |
| 41320 | 97 |
| 41320 | 233 |
| 41320 | 210  |
| 53011 | 178 |
| 53011 | 15 |
| 14123 | 108 |
| 14123 | 210 |
| 14123 | 97 |

> In the two table option, we only have to find "orane juice" in a single column, "item\_name", and it occurs once only (this is because this id number is a primary key, which means it can only occur once).

Author: ```suggestion
In the two table option, we only have to find "orane juice" in a single column, "item\_name", and it occurs once only (this is because this ID number is a primary key, which means it can only occur once).
```

> The theory of how to separate data for top speed and efficiency is complex. However, it's enough for now to understand that data in relational databases are fragmented in a process called normalization, to help reduce needless repetition, simplify data, and improve system speed and performance.

Reviewer: ```suggestion
The theory of how to separate data for top speed and efficiency is complex. However, it's enough for now to understand that data in relational databases are fragmented in a process called normalization, which 
- helps reduce needless repetition, 
- simplifies data, 
- and improves system speed and performance.
```

| 3411904329   | 320092     | 2020-01-30    | 56024     |
| 3411904330   | 612311     | 2020-01-31    | 99123     |

Which of the fields in this table is likely to be a primary key?  Select all that apply.

[[X]] med\_ord\_id
[[ ]] patient\_id
[[ ]] date
[[ ]] provider\_id
*********

<div class = "answer">

Because this table is about medication orders, and there's an identifier that looks like it uniquely identifies each of these interactions, we can guess that "med\_ord\_id" is probably a primary key.  On the other hand, we know for sure that "date" and "provider\_id" can't be primary keys, because they have repeating values.  That makes sense -- the same provider can write many different medication orders, and on any given date, there are likely to be many medication orders.  By that same logic, we also know that "patient\_id" can't be a primary key... it needs to be possible for a patient to have more than one medication order!

</div>

***********

Consider the same table.

Which of the fields in this table is likely to be a foreign key?  Select all that apply.

[[ ]] med\_ord\_id
[[X]] patient\_id
[[ ]] date
[[X]] provider\_id
*********

<div class = "answer">

> A foreign key is an identifier that originates in another table.  It certainly seems likely that there is a "patient" table and a "provider" table that the "patient\_id" and "provider\_id", respectively, originated in.  These are the two foreign keys.  The "med\_ord\_id" seems to originate here, in the "medication_order" table, so it's not a foreign key, and the "date" just seems like, well, a date, not an identifier at all!

Reviewer: Question: are there situations in which a date might be a primary key for say research on whether patients have better or worse outcomes on holidays or when new residents start in the summer? Or on rainy days?

This might be pushing too far, but I'm genuinely curious.

Author: yep, I could envision that, it's possible, but it's an edge case.
